domain/annotation/state/state.go

```diff
@@ -0,0 +1,283 @@
+// getAnnotationsForModel because for non-model ID Kinds we need to find the uuid of the id before
+// we retrieve annotations from the corresponding annotation table.
+func (st *State) getAnnotationsForID(ctx context.Context, db database.TxnRunner, id annotations.ID, getAnnotationsStmt *sqlair.Statement) (map[string]string, error) {
+	// Prepare queries for looking up the uuid of id
+	kindQuery, kindQueryParam, err := uuidQueryForID(id)
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", id.Name)
+	}
+	kindQueryStmt, err := sqlair.Prepare(kindQuery, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", id.Name)
+	}
+
+	// Running transactions for getting annotations
+	var annotationsResults []Annotation
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Looking up the UUID for id
+		result := sqlair.M{}
+		err = tx.Query(ctx, kindQueryStmt, kindQueryParam).Get(result)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Annotatef(err, "looking up UUID for ID: %s", id.Name)
+		}
+
+		if len(result) == 0 {
+			return fmt.Errorf("unable to find UUID for ID: %q %w", id.Name, errors.NotFound)
+		}
+
+		uuid, ok := result["uuid"].(string)
+		if !ok {
+			return fmt.Errorf("unable to find UUID for ID: %q %w", id.Name, errors.NotFound)
+		}
+		// Querying for annotations
+		return tx.Query(ctx, getAnnotationsStmt, sqlair.M{
+			"uuid": uuid,
+		}).GetAll(&annotationsResults)
+	})
+
+	if err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// No errors, we return empty map if no annotation is found
+			return map[string]string{}, nil
+		}
+		return nil, errors.Annotatef(err, "loading annotations for ID: %q", id.Name)
+	}
+
+	annotations := make(map[string]string, len(annotationsResults))
+
+	for _, result := range annotationsResults {
+		annotations[result.Key] = result.Value
+	}
+
+	return annotations, errors.Trace(err)
+}
+
+// SetAnnotations associates key/value annotation pairs with a given ID.
+// If annotation already exists for the given ID, then it will be updated with
+// the given value. Setting a key's value to "" will remove the key from the annotations map
+// (functionally unsetting the key).
+func (st *State) SetAnnotations(ctx context.Context, id annotations.ID,
+	annotationsParam map[string]string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Separate the annotations that are to be setted vs removed
+	toUpsert := make(map[string]string)
+	var toRemove []string
+
+	for key, value := range annotationsParam {
+		if value == "" {
+			toRemove = append(toRemove, key)
+		} else {
+			toUpsert[key] = value
+		}
+	}
+
+	// Prepare query (and parameters) for inserting and deleting annotations for id
+	setAnnotationsQuery := setAnnotationQueryForID(id)
+	toRemoveParamBindings, deleteAnnotationsQuerySqlairM := statement.SliceToSqlairMParams(toRemove)
```

Comment:
I don't think you need to jump through these hoops.

Seeing as individual annotations don't have UUIDs, you can just delete everything and set the whole map in a single transaction.

Reply:
I think deleting everything and inserting the new set might not be a good idea, because it means the caller then needs to provide the entire set of annotations at every call, even to update just one annotation. 

For example, let's say that I have `"foo":"4", "bar":"5", "baz":"6"`, and I wanna set `"baz":"3"`. Currently I can do that by calling `SetAnnotations` with just `"baz":"3"` (`foo` and `bar`'s values in the table won't be touched). But in the way you propose I have to call it with  `"foo":"4", "bar":"5", "baz":"3"`. 

The only advantage that I see in doing it your way (besides the simplicity in code), is that it would make the `unset` implicit (i.e. just don't provide that annotation in the input and it'll be gone - as opposed to explicitly setting it to `""`). But we'd lose the ability to talk about individual annotations and have to talk in terms of the whole set as I explained above.

@manadart @SimonRichardson wdyt?

Reply:
> because it means the caller then needs to provide the entire set of annotations at every call,

@manadart is right. This is how annotations works. You pass in the whole map, it's just a key/value store. I think we just delete all of them and add them in one swoop. It simplifies it, otherwise you need to read then work out the union (upserts), differences (removals and additions) and that seems a lot of busy work.